pyuvdata/utils.pyx

```diff
@@ -228,88 +227,114 @@ cpdef numpy.ndarray[dtype=numpy.float64_t] _ECEF_FROM_ENU(
     numpy.float64_t[::1] _lon,
     numpy.float64_t[::1] _alt,
 ):
-  cdef int i
-  cdef int nblts = enu.shape[0]
-  cdef numpy.ndarray[dtype=numpy.float64_t, ndim=2] xyz = np.empty((nblts, 3), dtype=np.float64)
-  cdef numpy.ndarray[dtype=numpy.float64_t, ndim=1] xyz_center = _xyz_from_latlonalt(_lat, _lon, _alt)
+    cdef Py_ssize_t i
+    cdef int nblts = enu.shape[1]
+    cdef numpy.float64_t sin_lat, cos_lat, sin_lon, cos_lon
 
-  # make a memoryview for the numpy array in c
-  cdef numpy.float64_t[:, ::1] _xyz = xyz
-  with nogil:
-    for i in range(nblts):
-      _xyz[i, 0] = (
-        - sin(_lat[0]) * cos(_lon[0]) * enu[i, 1]
-        - sin(_lon[0]) * enu[i, 0]
-        + cos(_lat[0]) * cos(_lon[0]) * enu[i, 2]
-        + xyz_center[0]
-      )
-      _xyz[i, 1] = (
-        - sin(_lat[0]) * sin(_lon[0]) * enu[i, 1]
-        + cos(_lon[0]) * enu[i, 0]
-        + cos(_lat[0]) * sin(_lon[0]) * enu[i, 2]
-        + xyz_center[1]
-      )
-      _xyz[i, 2] = cos(_lat[0]) * enu[i, 1] + sin(_lat[0]) * enu[i, 2] + xyz_center[2]
+    # allocate memory then make memory view for faster access
+    cdef numpy.ndarray[dtype=numpy.float64_t, ndim=2] _xyz = np.zeros((3, nblts), dtype=np.float64)
+    cdef numpy.float64_t[:, ::1] xyz = _xyz
 
-  return xyz
+    # we want a memoryview of the xyz of the center
+    # this looks a little silly but we don't have to define 2 different things
+    cdef numpy.float64_t[:] xyz_center = _xyz_from_latlonalt(_lat, _lon, _alt)[:, 0]
+
+    sin_lat = sin(_lat[0])
+    cos_lat = cos(_lat[0])
+
+    sin_lon = sin(_lon[0])
+    cos_lon = cos(_lon[0])
+
+    for i in range(nblts):
+        xyz[0, i] = (
+            - sin_lat * cos_lon * enu[1, i]
+            - sin_lon * enu[0, i]
+            + cos_lat * cos_lon * enu[2, i]
+            + xyz_center[0]
+        )
+        xyz[1, i] = (
+            - sin_lat * sin_lon * enu[1, i]
+            + cos_lon * enu[0, i]
+            + cos_lat * sin_lon * enu[2, i]
+            + xyz_center[1]
+        )
+        xyz[2, i] = cos_lat * enu[1, i] + sin_lat * enu[2, i] + xyz_center[2]
+
+    return _xyz
 
 # inital_uvw is a memoryviewed array as an input
 @cython.boundscheck(False)
 @cython.wraparound(False)
-cpdef numpy.ndarray[dtype=numpy.float64_t] _phase_uvw(
+cpdef numpy.ndarray[dtype=numpy.float64_t, ndim=2] _phase_uvw(
     numpy.float64_t ra,
```

Comment:
Should we be catching cases (either here or in `utils.py`) where `ra` and `dec` are passed as arrays? I know that doesn't make sense for how it's usually called, but we may want to sanitize the input (or at least make sure the user isn't doing anything unexpected).

Reply:
yeah that's a good point, I think we would need to do this in `utils.py`. I'll make sure it is sanitized against length 1 arrays, or scalars as arrays.

Reply:
So it looks like in `utils.py`, we expect inputs to be `float` type but don't do any explicit checking. I can add these checks in, but it will be a little more work for the multi-source branch to pull out later. Up to your discretion

Reply:
Let's not worry about it then. Probably not a huge deal, especially because in the docstring we explicitly say it's a `float` and not an `ndarray`.